backend/src/index.ts

```diff
@@ -0,0 +1,120 @@
+import express, { Request, Response } from 'express';
+import cors from 'cors';
+import { createUser, findUserByEmail, findUserById, findUserByUsername, updateUser } from './database';
+import { errorResponseBuilder, generateRandomPassword, isValidUser, parseUserForResponse } from './utils';
+
+const app = express();
+app.use(express.json());
+app.use(cors());
+
+app.post('/users/new', async (req: Request, res: Response) => {
+    const errorBuilder = errorResponseBuilder(res);
+
+    try {
+        const userData = req.body;
+
+        if (!isValidUser(req.body)) {
+            return errorBuilder.validationError();
+        }
+
+        const existingUserByEmail = await findUserByEmail(req.body.email);
```

Comment:
This endpoint alone is doing 3 requests to the DB. 

I am not a backend engineer and my knowledge around DB is very limited. But this looks suboptimal.
As the `create` endpoint will already answer with the right error codes you need. 
I would suggest to just call your `createUser` then building your API response from there.

`prisma.user.create` response will look like 

```
{
     code: "P2002",
     meta: { target: ["username"] },
}
``` 
`P2002` Being the "Unique constraint violation" according to Prisma's documentation.
And the target  Being `username` or `email` depending on which one is causing the violation.

Reply:
Excellent point, thank you for bringing it up!
I opened an issue just to remind myself: https://github.com/enekesabel/DDDForum/issues/4